design-proposals/klip-63-pause-resume-persistent-queries.md

```diff
@@ -0,0 +1,62 @@
+# KLIP-63: PAUSE and RESUME for persistent queries 
+
+**Author**: Jim Hughes (jnh5y) | 
+**Release Target**: Unknown | 
+**Status**: In Discussion | 
+**Discussion**: 
+
+**tl;dr:** Adds the ability to pause the processing of a persistent query.
+           
+## Motivation and background
+
+Users may wish to pause a query for a number of reasons and then resume processing.  Presently, they would have to terminate a query and recreate it.  This approach means that any state would have to be restored; the goal of this work is to add the capability to pause processing temporarily while allowing Kafka consumers and state stores to be retained.
+
+## What is in scope
+
+Pausing and resuming persistent queries and the necessary language modifications to support this are in scope.
+
+## What is not in scope
+
+We are not considering the ability to pause and resume transient queries.  Additionally, modifying the consumers' state is out of scope.  (This would allow for processing to skip records or reprocess data.)
```

Comment:
Even if we were to modify consumer's state, how would we do it? Were you thinking of something like Flink's savepoint which is a user-controlled checkpoint (which doesnt have an analogous equivalent in kafka streams)

Reply:
You've stated the problem with resuming from a different place (and hence why it is out of scope for this work!).  As far as I know, we don't have a clear way to do this in Streams.

In addition to savepoint, I've seen folks suggest managing the offsets for each partition.